### ISSUE TITLE
ACVP DRBG issue

### ISSUE THREAD

**Issue by Harshada-27** [usnistgov/ACVP#1432]
Vendor's DRBG is not testable via the current ACVP test framework at 256 bits strength.  From the current draft ([https://pages.nist.gov/ACVP/draft-vassilev-acvp-drbg.html#name-supported-values-per-drbg-o [pages.nist.gov]](https://urldefense.com/v3/__https:/pages.nist.gov/ACVP/draft-vassilev-acvp-drbg.html*name-supported-values-per-drbg-o__;Iw!!Ie6nUkZk!pqoUZwrlZZb1Y29dgjfmxpzQYJHYBJbxz3a51lBNX2HhNLIyi9e9BukZDj9kyaRzKW3jZMaCIKMa5kMr1gtr7UmERFAGlH-FkA$)), for cntrDBG, derFunc=true, AES-256, the min entropy size is 256 bits and min nonce is 128 bits.  Vendor's implementation uses 440 bits of entropy and a 72 bit nonce.  As the initialization concatenates the 1st entropy value + nonce, the distinction between entropy and nonce is artificial, what really counts is the total number of bits.  As 256 + 128 = 384 bits, and 440 + 72 = 512 bits, vendor do supply enough bits for the initialization.
For AES-192 we use 312 bits entropy and a 128 bit nonce, above the minimums of 192 and 96 respectively.  For AES-128 we use 184 bits entropy and a 128 bit nonce, above the minimums of 128 and 64 respectively.

Could the CAVP please advise if the nonce length can be customized in the request files? At present the server seems to be sending default values which do not seem to work for the vendor’s implementation.

**Comment by jbrock24** [comment]
If I'm understanding you correctly, you can do that in the [registration file](https://pages.nist.gov/ACVP/draft-vassilev-acvp-drbg.html#name-drbg-registration-example):

```
"capabilities": [
    {
      "mode": "AES-128",
      "derFuncEnabled": false,
      "entropyInputLen": [
        256
      ],
      "nonceLen": [
        256
      ],
      "persoStringLen": [
        256
      ],
      "additionalInputLen": [
        256
      ],
      "returnedBitsLen": 512
    }
  ]
```

**Comment by jbrock24** [comment]
[Here](https://pages.nist.gov/ACVP/draft-vassilev-acvp-drbg.html#supported_values) are the values we support for that. Are you are talking about an actual _custom_ length for your own use outside of the Support Values in the base system?

**Comment by Harshada-27** [comment]
Yes, correct. For AES-256, the min nonce length defined in the ACVTS documentation/specification is 128 bits, whereas the vendor supports a 72-bit min nonce. Could the CAVP please assist/guide as to how this can be tested?

**Comment by livebe01** [comment]
@Harshada-27 can you supply the complete json registration for the IUT you wish to test?

**Comment by jbrock24** [comment]
Hi @Harshada-27 - Just wondering if this is still something you will need?

**Comment by avaralikar** [comment]
Hi @jbrock24, Acumen Security would like to please re-open this topic. Thank you for the following feedback:

To accomplish your testing, adjust your registration properties such that ACVTS provides you test cases such that the size of entropyInput + nonce = the size of your entropy + nonce and then have your test harness make the break at your 440 + 72 boundary.

Please see the vendor's input inline below:
If we adjust the size of entropy input and nonce in the registration, the very first input to the DRBG will be the correct size, but after that every entropyInput value will be the wrong size – so the test will not pass. 


Since this is needed as part of a FIPS validation, we have also notified the CMVP who have advised us to further explore this here. We wanted to thus request your guidance and would greatly appreciate the same.

**Comment by livebe01** [comment]
Hi @avaralikar, thank you for reopening the communication regarding this issue. I will review the vendor's response which you've provided and get back to you with a response/way forward in the next week.

**Comment by msonje** [comment]
Good Morning @jbrock24, 
We wanted to check is there any feedback on this topic. Since this is needed as part of a FIPS validation and we can plan our CMVP submission accordingly.

**Comment by livebe01** [comment]
I wanted to provide an update on the issue raised in this ticket. SP 800-90Ar1 page 27 contains the following statement.

> Note that if a random value is used in the
nonce, the entropy_input and random portion of the nonce could be acquired using a
single Get_entropy_input call (see step 6 of the instantiate process); in this case, the
first parameter of the Get_entropy_input call is adjusted to include the entropy for the
nonce (i.e., the security_strength is increased by at least ½ security_strength, and min-
length is increased to accommodate the length of the nonce), instantiate process step 8 is
omitted, and the nonce is omitted from the parameter list in instantiate process step 9.

It's clear that the ACVTS DRBG testing needs to be updated to support smaller, even 0-sized nonce lengths. We plan to include updated DRBG testing in our next release. We'd like to have that release deployed to our Demo server for testing by 1/26.

**Comment by msonje** [comment]
Thank you so much, @livebe01, for the information and development on the Demo server. We will try this out on the demo. Is there any ETA for when we can test the DRBG on the prod server as well?

**Comment by livebe01** [comment]
Sure, we'd like to get the v1.1.0.33 release onto Demo by Wednesday of this week. Assuming we don't have any issues with v1.1.0.33 on Demo, we'd like to send it to Prod the week after, perhaps on the 7th.

**Comment by livebe01** [comment]
The fix for this is on Demo in release [v1.1.0.33](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.33)

**Comment by msonje** [comment]
Thank you for the update.

**Comment by livebe01** [comment]
The fix for this is on Prod in release [v1.1.0.33](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.33)
